src/analysis/normed_space/hahn_banach/separation.lean

```diff
@@ -0,0 +1,350 @@
+* `geometric_hahn_banach_open`: One set is open. Weak separation.
+* `geometric_hahn_banach_open_point`, `geometric_hahn_banach_point_open`: One set is open, the
+  other is a singleton. Weak separation.
+* `geometric_hahn_banach_open_open`: Both sets are open. Semistrict separation.
+* `geometric_hahn_banach_compact_closed`, `geometric_hahn_banach_closed_compact`: One set is closed,
+  the other one is compact. Strict separation.
+* `geometric_hahn_banach_point_closed`, `geometric_hahn_banach_closed_point`: One set is closed, the
+  other one is a singleton. Strict separation.
+* `geometric_hahn_banach_point_point`: Both sets are singletons. Strict separation.
+-/
+
+open set
+open_locale pointwise
+
+variables {𝕜 E : Type*}
+
+lemma continuous_at_of_exists_open [normed_ring 𝕜] [normed_group E] [module 𝕜 E] (f : E →ₗ[𝕜] 𝕜)
+  (hf : ∀ ε, 0 < ε → ∃ (U : set E), (0 : E) ∈ U ∧ is_open U ∧ ∀ y ∈ U, ∥f y∥ < ε) :
+  continuous_at f 0 :=
+begin
+  intros U hU,
+  rw metric.nhds_basis_ball.1 at hU,
+  rcases hU with ⟨ε, hε₁, hε₂⟩,
+  simp only [filter.mem_map],
+  obtain ⟨V, hV₁, hV₂, hV₃⟩ := hf ε hε₁,
+  rw mem_nhds_iff,
+  refine ⟨V, λ y hy, hε₂ _, hV₂, hV₁⟩,
+  rw [metric.mem_ball, f.map_zero, dist_zero_right],
+  exact hV₃ _ hy,
+end
+
+lemma continuous_at_of_exists_open' [normed_ring 𝕜] [normed_group E] [module 𝕜 E]
+  (f : E →ₗ[𝕜] 𝕜) {x : E}
+  (hf : ∀ ε, 0 < ε → ∃ (U : set E), x ∈ U ∧ is_open U ∧ ∀ y ∈ U, ∥f y - f x∥ < ε) :
+  continuous_at f x :=
+begin
+  intros U hU,
+  rw metric.nhds_basis_ball.1 at hU,
+  rcases hU with ⟨ε, hε₁, hε₂⟩,
+  simp only [filter.mem_map],
+  obtain ⟨V, hV₁, hV₂, hV₃⟩ := hf ε hε₁,
+  rw mem_nhds_iff,
+  refine ⟨V, λ y hy, hε₂ _, hV₂, hV₁⟩,
+  rw [metric.mem_ball, dist_eq_norm],
+  exact hV₃ _ hy,
+end
+
+/-- A nonzero continuous linear functional is open. -/
+lemma nonzero_linear_map_is_open_map [topological_space 𝕜] [division_ring 𝕜]
+  [topological_ring 𝕜] [add_comm_group E] [topological_space E] [topological_add_group E]
+  [module 𝕜 E] [has_continuous_smul 𝕜 E] (f : E →L[𝕜] 𝕜) (hf : f ≠ 0) :
+  is_open_map f :=
+begin
+  obtain ⟨x₀, hx₀⟩ : ∃ x₀, f x₀ ≠ 0,
+  { by_contra h,
+    push_neg at h,
+    exact hf (continuous_linear_map.ext (λ x, by simp [h]) )},
+  intros A hA,
+  rw is_open_iff_mem_nhds,
+  rintro _ ⟨a, ha, rfl⟩,
+  let g : 𝕜 → E := λ x, a + (x - f a) • (f x₀)⁻¹ • x₀,
+  have := (show continuous g, by continuity).is_open_preimage _ ‹is_open A›,
+  rw is_open_iff_mem_nhds at this,
+  exact filter.sets_of_superset _ (this (f a) (by simpa [set.mem_preimage, g]))
+    (λ x hx, ⟨_, hx, by simp [hx₀]⟩),
+end
+
+variables [normed_group E] [normed_space ℝ E]
+
+/-- Given a set `C` which is a convex neighbourhood of `0` and a point `x₀` outside of it, there is
+a continuous linear functional `f` separating `x₀` and `C`, in the sense that it sends `x₀` to 1 and
+all of `C` to values strictly below `1`. -/
+lemma separate_convex_open_set {C : set E} (zero_mem : (0:E) ∈ C) (hC : convex ℝ C)
```

Comment:
I did some golf and extracted a lemma from that proof:
```lean
lemma gauge_lt_of_mem_smul (x : E) (ε : ℝ) (hε : 0 < ε) {C : set E} (zero_mem : (0:E) ∈ C)
  (hC : convex ℝ C) (hC₂ : is_open C) (hx : x ∈ ε • C) :
  gauge C x < ε :=
begin
  have : ε⁻¹ • x ∈ C,
  { rwa ←mem_smul_set_iff_inv_smul_mem₀ hε.ne' },
  have h_gauge_lt := gauge_lt_one_of_mem_of_open hC zero_mem hC₂ (ε⁻¹ • x) this,
  rwa [gauge_smul_of_nonneg (inv_nonneg.2 hε.le), smul_eq_mul, inv_mul_lt_iff hε, mul_one]
    at h_gauge_lt,
  apply_instance
end

/-- Given a set `C` which is a convex neighbourhood of `0` and a point `x₀` outside of it, there is
a continuous linear functional `f` separating `x0` and `C`, in the sense that it sends `x₀` to 1 and
all of `C` to values strictly below `1`. -/
lemma separate_convex_open_set {C : set E} (zero_mem : (0:E) ∈ C) (hC : convex ℝ C)
  (hC₂ : is_open C) {x₀ : E} (hx₀ : x₀ ∉ C) :
  ∃ (f : E →L[ℝ] ℝ), f x₀ = 1 ∧ ∀ x ∈ C, f x < 1 :=
begin
  let f : linear_pmap ℝ E ℝ :=
    linear_pmap.mk_span_singleton x₀ 1 (ne_of_mem_of_not_mem zero_mem hx₀).symm,
  have hfx₀ : f ⟨(1:ℝ) • x₀, by { dsimp, rw submodule.mem_span_singleton, refine ⟨1, rfl⟩ }⟩ = 1,
    by simp_rw [linear_pmap.mk_span_singleton_apply, one_smul],
  rcases exists_extension_of_le_sublinear f (gauge C) _ _ _ with ⟨φ, hφ₁, hφ₂⟩,
  { refine ⟨⟨φ, (φ.to_add_monoid_hom.uniform_continuous_of_continuous_at_zero _).continuous⟩, _, _⟩,
    { refine continuous_at_of_exists_open _ (λ ε hε, ⟨(ε • C) ∩ (-ε • C), ⟨_, _⟩, _, _⟩),
      { exact mem_smul_set.mpr ⟨0, zero_mem, smul_zero _⟩ },
      { exact mem_smul_set.mpr ⟨0, zero_mem, smul_zero _⟩ },
      { exact (is_open_map_smul₀ hε.ne' _ hC₂).inter
          (is_open_map_smul₀ (neg_ne_zero.mpr hε.ne.symm) _ hC₂), },
      rintro x ⟨hx₁, hx₂⟩,
      rw [real.norm_eq_abs, abs_lt, neg_lt, ←linear_map.map_neg],
      split; apply (hφ₂ _).trans_lt,
      { refine gauge_lt_of_mem_smul (-x) ε hε zero_mem hC hC₂ _,
        rw [mem_smul_set_iff_inv_smul_mem₀ hε.ne', smul_neg],
        rwa [mem_smul_set_iff_inv_smul_mem₀ (neg_ne_zero.mpr hε.ne'), inv_neg, neg_smul] at hx₂, },
      { exact gauge_lt_of_mem_smul x ε hε zero_mem hC hC₂ hx₁, } },
    { dsimp,
      have : x₀ ∈ f.domain := submodule.mem_span_singleton_self _,
      rw [←submodule.coe_mk x₀ this, hφ₁, ← hfx₀],
      congr,
      rw one_smul, },
    { exact λ x hx, (hφ₂ x).trans_lt (gauge_lt_one_of_mem_of_open hC zero_mem hC₂ _ hx) } },
  { simp_rw ← smul_eq_mul,
    exact λ c hc x, gauge_smul_of_nonneg hc.le x, },
  { exact gauge_add_le hC (absorbent_nhds_zero (hC₂.mem_nhds zero_mem)) },
  { rintro ⟨x, hx⟩,
    obtain ⟨y, rfl⟩ := submodule.mem_span_singleton.1 hx,
    rw linear_pmap.mk_span_singleton_apply,
    simp only [mul_one, algebra.id.smul_eq_mul, submodule.coe_mk],
    cases lt_or_le 0 y with h h,
    { rw [gauge_smul_of_nonneg h.le, smul_eq_mul, le_mul_iff_one_le_right h],
      exact one_le_gauge_of_not_mem hC zero_mem hC₂ hx₀,
      apply_instance },
    { exact h.trans (gauge_nonneg _) } }
end
```

Reply:
Oh thanks!